### PR TITLE
Change centered-container to use max width and allow nesting of classes

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Change the way the `.centered-container` works. The current solution only works at the outer most container element, often children of `body` or such. Otherwise you get gutters inside gutters inside gutters. This change allows the user to nest the classes and properly set the element to have appropriate gutters at all breakpoints.

ex: 
```
<div class="cc">
   <div class="cc">
    <div class="cc">
       this element would be 450px inset from the window on each side while on 2xl screens, and 75px on each side on mobile. The proposed change would make sure this text would only have 25px on mobile and 150px on 2xl.
    </div>
  </div>
</div>
```

### CODE REVIEW

File: src/plugins/components/layout/centered-container.js [hunk 1]
```diff
@@ -2,36 +2,46 @@
  * Centered container.
  */
 module.exports = function () {
-  return function ({ addComponents, theme }) {
+  return function ({ addUtilities, theme }) {
```
Comment (sherakama) [inline src/plugins/components/layout/centered-container.js:5]: Why the change to Utility here?

File: src/plugins/components/layout/centered-container.js [hunk 2]
```diff
@@ -2,36 +2,46 @@
...
     // Find and set the padding based on the screen margins
     const screens = theme('screens');
-    const padding = {};
+    const maxWdiths = {};
+    let largestScreen = 0;
+    let largestGutter = 0;
+    let smallestGutter = 999999;
```
Comment (sherakama) [inline src/plugins/components/layout/centered-container.js:11]: Just shy of a million?

File: src/plugins/components/layout/centered-container.js [hunk 3]
```diff
@@ -2,36 +2,46 @@
...
+      }
+    });
+      
+    const key = `@media only screen and (min-width: ${largestScreen + (largestGutter * 2)}px)`
     const components = {
       // Center an element horizontally.
-      '.centered-container, .cc': {
-        paddingLeft: theme('decanter.screenMargins.xs'),
-        paddingRight: theme('decanter.screenMargins.xs'),
+      '.centered': {
```
Comment (sherakama) [inline src/plugins/components/layout/centered-container.js:33]: Why the change in name?

**Comment by sherakama** [comment]
I like this change in behavior as it matches my expectation of how it should behave and how other CSS libs behave. A couple of questions in here first.

File: src/plugins/components/layout/centered-container.js (hunk 1 above)
Comment (pookmish) [reply src/plugins/components/layout/centered-container.js:5]: > Use the addUtilities function to register simple static utilities that don’t support user-provided values

My belief is this matches the criteria. But I honestly think they accomplish the same things. :shrug:

File: src/plugins/components/layout/centered-container.js (hunk 2 above)
Comment (pookmish) [reply src/plugins/components/layout/centered-container.js:11]: dr evil

File: src/plugins/components/layout/centered-container.js [hunk 4]
```diff
@@ -2,36 +2,46 @@
...
-        paddingLeft: theme(`decanter.screenMargins.${key}`),
-        paddingRight: theme(`decanter.screenMargins.${key}`),
-      };
-    });
+      if (largestScreen < parseInt(screens[key])) largestScreen = parseInt(screens[key]);
+
+      if (theme(`decanter.screenMargins.${key}`)) {
+        const gutterWidth = parseInt(theme(`decanter.screenMargins.${key}`));
+        if (largestGutter < gutterWidth) largestGutter = gutterWidth;
+        if (smallestGutter > gutterWidth) smallestGutter = gutterWidth;
```
[the remark this answers is not in the document]
Comment (pookmish) [reply src/plugins/components/layout/centered-container.js:22]: well we don't know which one is the smallest or in which order they are coming in. I think its best to allow for as dynamic as it can be.

File: src/plugins/components/layout/centered-container.js (hunk 3 above)
Comment (pookmish) [reply src/plugins/components/layout/centered-container.js:33]: Woops... that's a copy pasta error.

File: src/plugins/components/layout/centered-container.js (hunk 1 above)
Comment (sherakama) [reply src/plugins/components/layout/centered-container.js:5]: https://tailwindcss.com/docs/adding-custom-styles

```
The base layer is for things like reset rules or default styles applied to plain HTML elements.
The components layer is for class-based styles that you want to be able to override with utilities.
The utilities layer is for small, single-purpose classes that should always take precedence over any other styles.
```

I believe we chose the components option for this one so that other utility classes would win if the developer chose to add different behaviours later.

Comment (pookmish) [reply src/plugins/components/layout/centered-container.js:5]: ahh. interesting! yeah i agree. I didn't realize that it would add them at different layer

**Comment by pookmish** [comment]
I did this implementation and it's so much easier to just set the class on the drupal paragraph and I never have to worry about the container creating padding or what it's size or left/right side contains.

**Comment by yvonnetangsu** [comment]
@pookmish @sherakama Was in meeting this morning so I need to catch up with the full thread of conversations, but, we did have a reason that we used the padding method for the existing `su-cc` class - it's a long story due to having it exactly match the Decanter v6 `centered-container` mixin and it's because of UComm's request for that approach.

Basically, using existing `su-cc`, it would allow one to do this
```
<div class="su-cc su-bg-digital-red">
Some text
</div>
```

It will create the max width container around the text, but still have the background color shown as edge to edge. UComm wanted to use as few nested container as possible so that's why we did it that way (instead of using outer container with bg color, and inner container with su-cc).

So, because of that history, directly modify the existing `su-cc` class would break backward compatibility for our sites 😬 . Are you open to adding your approach as `su-cc-alt` perhaps @pookmish?

**Comment by yvonnetangsu** [comment]
@pookmish Personally I'd use your approach too if it's not because of backward compatibility and the UComm request back then.
cc @sherakama

**Comment by pookmish** [comment]
> So, because of that history, directly modify the existing `su-cc` class would break backward compatibility for our sites grimacing . Are you open to adding your approach as `su-cc-alt` perhaps @pookmish?

Yet again, poor implementation for an obscure use with a single site due to legacy issues. Feel free to do what you wish to my PR. I'll just keep overridding more and more or using less of Decanter in my projects.

**Comment by yvonnetangsu** [comment]
> > So, because of that history, directly modify the existing `su-cc` class would break backward compatibility for our sites grimacing . Are you open to adding your approach as `su-cc-alt` perhaps @pookmish?
> 
> Yet again, poor implementation for an obscure use with a single site due to legacy issues. Feel free to do what you wish to my PR. I'll just keep overridding more and more or using less of Decanter in my projects.

It's not a single site. We have half a dozen live sites using decanter v7 and making this change would require us to test all those sites and make changes if needed.

**Comment by sherakama** [comment]
> Yet again, poor implementation for an obscure use with a single site due to legacy issues. Feel free to do what you wish to my PR. I'll just keep overridding more and more or using less of Decanter in my projects.

@pookmish here's your tiny violin. 



@yvonnetangsu, This sounds like we have two different use cases here and we have a need for a centered banner and a centered chromeless. And to add both, we would have a naming issue. the centered container should be the chromeless option IMHO but we should support the banner too.

Can we do both and come up with class names that make sense? I'm ok for a find-all-and-replace type change.

**Comment by pookmish** [comment]
Ok, I won't be making contributions any longer. There's too much difficulty to make improvements.

**Comment by sherakama** [comment]
@pookmish 

Thank you for meeting with me and talking this through. I apologize again for the inappropriate giphy but I am hoping you can write out a few answers to some of the questions we talked about and a couple of new ones to document your challenges. 

* What would you have liked the outcome of this PR to be if straight acceptance was not possible or if a disagreement on the legitimacy of existing use cases?
* How does the suggested compromise in this thread add technical debt? How does it negatively affect your work?
* How does "overridding more and more or using less of Decanter" negatively affect your work?
* What are your tradeoffs to abandoning Decanter? ie: what do you stand to gain/lose?

You also mentioned that overriding individual class names isn't possible and that it creates a merge of styles. I've flagged that as something to look into. 

thanks,

**Comment by jbickar** [comment]
> @pookmish
> 
> Thank you for meeting with me and talking this through. I apologize again for the inappropriate giphy but I am hoping you can write out a few answers to some of the questions we talked about and a couple of new ones to document your challenges.
> 
>     * What would you have liked the outcome of this PR to be if straight acceptance was not possible or if a disagreement on the legitimacy of existing use cases?
> 
>     * How does the suggested compromise in this thread add technical debt? How does it negatively affect your work?
> 
>     * How does "overridding more and more or using less of Decanter" negatively affect your work?
> 
>     * What are your tradeoffs to abandoning Decanter? ie: what do you stand to gain/lose?
> 
> 
> You also mentioned that overriding individual class names isn't possible and that it creates a merge of styles. I've flagged that as something to look into.
> 
> thanks,

@pookmish, please respond in this thread.

**Comment by sherakama** [comment]
Hey @pookmish,

We'd like to wrap this up with you and need your response. Please provide your response to the final questions and we can close this out and move along.

**Comment by sherakama** [comment]
@pookmish It's been more than two weeks and I haven't heard back from you. Please wrap this up by answering the remaining questions. Thank you.

**Comment by sherakama** [comment]
@pookmish this is a reminder about the thread you agreed to answer when we met in our 1:1 on Tuesday. Please answer the questions in https://github.com/SU-SWS/decanter/pull/911#issuecomment-1561620893
